sdk/testing/stepwise/environments/docker/environment.go

```diff
@@ -0,0 +1,851 @@
+	})
+
+	n.ServerCertPEMFile = filepath.Join(n.WorkDir, "cert.pem")
+	err = ioutil.WriteFile(n.ServerCertPEMFile, n.ServerCertPEM, 0755)
+	if err != nil {
+		return err
+	}
+
+	n.ServerKeyPEMFile = filepath.Join(n.WorkDir, "key.pem")
+	err = ioutil.WriteFile(n.ServerKeyPEMFile, n.ServerKeyPEM, 0755)
+	if err != nil {
+		return err
+	}
+
+	tlsCert, err := tls.X509KeyPair(n.ServerCertPEM, n.ServerKeyPEM)
+	if err != nil {
+		return err
+	}
+
+	certGetter := reloadutil.NewCertificateGetter(n.ServerCertPEMFile, n.ServerKeyPEMFile, "")
+	if err := certGetter.Reload(); err != nil {
+		return err
+	}
+	tlsConfig := &tls.Config{
+		Certificates:   []tls.Certificate{tlsCert},
+		RootCAs:        n.Cluster.RootCAs,
+		ClientCAs:      n.Cluster.RootCAs,
+		ClientAuth:     tls.RequestClientCert,
+		NextProtos:     []string{"h2", "http/1.1"},
+		GetCertificate: certGetter.GetCertificate,
+	}
+
+	if n.Cluster.clientAuthRequired {
+		tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
+	}
+	n.TLSConfig = tlsConfig
+
+	return nil
+}
+
+// DockerClusterNode represents a single instance of Vault in a cluster
+type DockerClusterNode struct {
+	NodeID            string
+	Address           *net.TCPAddr
+	HostPort          string
+	Client            *api.Client
+	ServerCert        *x509.Certificate
+	ServerCertBytes   []byte
+	ServerCertPEM     []byte
+	ServerCertPEMFile string
+	ServerKey         *ecdsa.PrivateKey
+	ServerKeyPEM      []byte
+	ServerKeyPEMFile  string
+	TLSConfig         *tls.Config
+	WorkDir           string
+	Cluster           *DockerCluster
+	container         *types.ContainerJSON
+	dockerAPI         *docker.Client
+}
+
+// CreateAPIClient creates and configures a Vault API client to communicate with
+// the running Vault Cluster for this DockerClusterNode
+func (n *DockerClusterNode) CreateAPIClient() (*api.Client, error) {
+	transport := cleanhttp.DefaultPooledTransport()
+	transport.TLSClientConfig = n.TLSConfig.Clone()
+	if err := http2.ConfigureTransport(transport); err != nil {
+		return nil, err
+	}
+	client := &http.Client{
+		Transport: transport,
+		CheckRedirect: func(*http.Request, []*http.Request) error {
+			// This can of course be overridden per-test by using its own client
+			return fmt.Errorf("redirects not allowed in these tests")
+		},
+	}
+	config := api.DefaultConfig()
+	if config.Error != nil {
+		return nil, config.Error
+	}
+	config.Address = fmt.Sprintf("https://127.0.0.1:%s", n.HostPort)
+	config.HttpClient = client
+	config.MaxRetries = 0
+	apiClient, err := api.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+	apiClient.SetToken(n.Cluster.RootToken())
+	return apiClient, nil
+}
+
+// Cleanup kills the container of the node
+func (n *DockerClusterNode) Cleanup() error {
+	return n.dockerAPI.ContainerKill(context.Background(), n.container.ID, "KILL")
+}
+
+func (n *DockerClusterNode) Start(cli *docker.Client, caDir, netName string, netCIDR *DockerClusterNode, pluginBinPath string) error {
```

Comment:
Since this is not called outside of the package can we make this private?

Reply:
On a second thought, I wonder if `DockerClusterNode` can be made private.

Reply:
I made the `Start` and `DockerClusterNode` private in 863f2a1ba